File: event-bus-rabbit/src/main/java/pegasus/topology/service/StaticTopologyManager.java
package pegasus.topology.service;

import java.util.HashMap;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import pegasus.eventbus.amqp.RoutingInfo;
import pegasus.eventbus.client.EventManager;
import pegasus.topology.event.RegisterClient;
import pegasus.topology.event.UnregisterClient;
import pegasus.topology.event.TopologyUpdate;

public class StaticTopologyManager implements TopologyManager {

    protected static final Logger      LOG                       = LoggerFactory.getLogger(StaticTopologyManager.class);
    private static final String        DEFAULT_TOPOLOGY_EXCHANGE = "topology";

    private String                     topologyExchange          = DEFAULT_TOPOLOGY_EXCHANGE;
    private Map<String, RoutingInfo>   topologyEventRegistry     = new HashMap<String, RoutingInfo>();
    private Map<String, RoutingInfo[]> topologyEventSetRegistry  = new HashMap<String, RoutingInfo[]>();

    public StaticTopologyManager() {
        this.topologyExchange = DEFAULT_TOPOLOGY_EXCHANGE;
        initializeTopologyRegistries();
    }

    public StaticTopologyManager(String topologyExchange) {
        this.topologyExchange = topologyExchange;
        initializeTopologyRegistries();
    }

    private void initializeTopologyRegistries() {
        String registerClientTopic = RegisterClient.class.getCanonicalName();
        RoutingInfo registerClientRoute = new RoutingInfo(topologyExchange, RoutingInfo.ExchangeType.Topic, true, registerClientTopic);
        topologyEventRegistry.put(registerClientTopic, registerClientRoute);
        String unregisterClientTopic = UnregisterClient.class.getCanonicalName();
        RoutingInfo unregisterClientRoute = new RoutingInfo(topologyExchange, RoutingInfo.ExchangeType.Topic, true, unregisterClientTopic);
        topologyEventRegistry.put(unregisterClientTopic, unregisterClientRoute);
        String topologyUpdateTopic = TopologyUpdate.class.getCanonicalName();
        RoutingInfo topologyUpdateRoute = new RoutingInfo(topologyExchange, RoutingInfo.ExchangeType.Topic, true, topologyUpdateTopic);
        topologyEventRegistry.put(topologyUpdateTopic, topologyUpdateRoute);

        RoutingInfo[] allRoutes = { new RoutingInfo(topologyExchange, RoutingInfo.ExchangeType.Topic, true, "#") };
        topologyEventSetRegistry.put("ALL", allRoutes);
    }

    @Override
    public RoutingInfo getRoutingInfoForEvent(Class<?> eventType) {

        LOG.info("Looking for route for event type [{}] in static topology mapper.", eventType.getCanonicalName());

        RoutingInfo route = null;
        String topic = eventType.getCanonicalName();
        if (topologyEventRegistry.containsKey(topic)) {
            route = topologyEventRegistry.get(topic);

            LOG.info("Found route [{}] in static topology mapper.", route);
        }
        return route;
    }

    @Override
    public RoutingInfo[] getRoutingInfoForNamedEventSet(String eventSetName) {

        LOG.info("Looking for routes for event set name [{}] in static topology mapper.", eventSetName);

        RoutingInfo[] routes = null;
        if (topologyEventSetRegistry.containsKey(eventSetName)) {
            routes = topologyEventSetRegistry.get(eventSetName);

            LOG.info("Found routes [{}] in static topology mapper.", routes);
        }
        return routes;
    }

    @Override
    public void start(EventManager eventManager) {
        // ignore
    }

    @Override
    public void stop() {
        // ignore
    }

}

File: event-bus-rabbit/src/main/dotnet/topology/StaticTopologyManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using log4net;

using pegasus.eventbus.client;
using pegasus.eventbus.amqp;


namespace pegasus.eventbus.topology
{
    public class StaticTopologyManager : ITopologyService
    {
        private static readonly ILog LOG = LogManager.GetLogger(typeof(StaticTopologyManager));
        private static readonly string DEFAULT_TOPOLOGY_EXCHANGE = "topology";

        private string _topologyExchange;
        private IDictionary<string, RoutingInfo> _topologyEventRegistry;
        private IDictionary<string, IEnumerable<RoutingInfo>> _topologyEventSetRegistry;


        public StaticTopologyManager()
        {
            _topologyExchange = DEFAULT_TOPOLOGY_EXCHANGE;
            _topologyEventRegistry = new Dictionary<string, RoutingInfo>();
            _topologyEventSetRegistry = new Dictionary<string, IEnumerable<RoutingInfo>>();

            this.RegisterForEvents();
        }


        public void Start(IEventManager eventManager)
        {
        }

        public void Close()
        {
        }

        public RoutingInfo GetRoutingInfoForEventOfType(Type evType)
        {
            RoutingInfo route = null;

            if (_topologyEventRegistry.ContainsKey(evType.FullName))
            {
                route = _topologyEventRegistry[evType.FullName];
            }

            return route;
        }

        public IEnumerable<RoutingInfo> GetRoutingInfoForNamedEventSet(string eventSetName)
        {
            // the static topo service never adds to the event set dictionary,
            // so instead of looking it up and having it always return null,
            // I'm just going to always return null
            return null;
        }


        private void RegisterForEvents()
        {
            this.RegisterForEvent(typeof(RegisterClient));
            this.RegisterForEvent(typeof(UnregisterClient));
    	    this.RegisterForEvent(typeof(HeartBeat));
      	    this.RegisterForEvent(typeof(TopologyUpdate));
     	    this.RegisterForEvent(typeof(GetEventTypeRoute));
            this.RegisterForEvent(typeof(EventTypeRoutingInfo));
        }

        private void RegisterForEvent(Type eventType)
        {
            _topologyEventRegistry.Add(
                eventType.FullName, 
                new RoutingInfo(_topologyExchange, ExchangeType.Topic, true, eventType.FullName));
        }
    }
}
